examples/classifier.c

```diff
@@ -15,6 +16,324 @@ float *get_regression_values(char **labels, int n)
+        free_data(train);
+
+        if(*net->seen/N > epoch) {
+            epoch = *net->seen/N;
+            update_epoch = 1;
+        }
+
+        if(update_epoch && (epoch % params.eval_epochs) == 0){
+            char buff[1024];
+            sprintf(buff, "%s/%s.weights",backup_directory, base);
+            save_weights(net, buff);
+            update_epoch = 0;
+
+            int cur_epoch = epoch / params.eval_epochs;
+
+            //train_accs[cur_epoch] = calculate_accurracy(datacfg, cfgfile, buff, "train");
+            //valid_accs[cur_epoch] = calculate_accurracy(datacfg, cfgfile, buff, "valid");
+
+            float** conf_train = calculate_confusion_matrix(datacfg, cfgfile, buff, "train");
+            float** conf_valid = calculate_confusion_matrix(datacfg, cfgfile, buff, "valid");
+
+            train_accs[cur_epoch] = eval(conf_train, classes);
+            valid_accs[cur_epoch] = eval(conf_valid, classes);
+
+            for(j = 0; j < classes; ++j)
+                free(conf_train[j]), free(conf_valid[j]);
+            free(conf_train), free(conf_valid);
+
+
+            printf("\n\nTrain metric: %f\nValidation metric: %f\n\n\n", train_accs[cur_epoch], valid_accs[cur_epoch]);
+
+            fprintf(training_log, "%f %f\n", train_accs[cur_epoch], valid_accs[cur_epoch]);
+
+            if(valid_accs[cur_epoch] > best_acc){
+                best_acc = valid_accs[cur_epoch];
+                sprintf(buff, "%s/%s-best.weights",backup_directory,base);
+                save_weights(net, buff);
+            }
+
+            if(cur_epoch > 0 && valid_accs[cur_epoch] > valid_accs[cur_epoch-1])
+                bad_epochs = 0;
+
+            else if(++bad_epochs > params.patience)
+                break;
+        }
+    }
+
+    char buff[1024];
+    sprintf(buff, "%s/%s.weights", backup_directory, base);
+    save_weights(net, buff);
+    pthread_join(load_thread, 0);
+
+    free_network(net);
+    if(labels) free_ptrs((void**)labels, classes);
+    free_ptrs((void**)paths, plist->size);
+    free_list(plist);
+    //free(valid_accs);
+    //free(train_accs);
+    fclose(training_log);
+    //free(base);
+}
+
+
+
+
+float** calculate_confusion_matrix (char *datacfg, char *filename, char *weightfile, char* set_name)
+{
+    int i, j;
+    network *net = load_network(filename, weightfile, 0);
+    set_batch_network(net, 1);
+    srand(time(0));
+
+    list *options = read_data_cfg(datacfg);
+
+    char *label_list = option_find_str(options, "labels", "data/labels.list");
+    char *leaf_list = option_find_str(options, "leaves", 0);
+    if(leaf_list) change_leaves(net->hierarchy, leaf_list);
+    char *valid_list = option_find_str(options, set_name, "data/test.list");
+    int classes = option_find_int(options, "classes", 2);
+    int topk = option_find_int(options, "top", 1);
+
+    char **labels = get_labels(label_list);
+    list *plist = get_paths(valid_list);
+
+    char **paths = (char **)list_to_array(plist);
```

Comment:
Rename `paths` to something that better reflect its meaning, such as `validation_images_paths`